fix: make_filter_cmd for powershell as shell

3N4N opened the pull request:
Closes #15913  which prevented nvim from creating out temp file for `:!cmd` commands over a range of lines. The reason is because PowerShell doesn't use `<` as the stdin redirection. Instead, it uses a parameter `-RedirectStandardInput`. The rest is explained in the bug report.

3N4N committed:
fix: make_filter_cmd for :! powershell #15913

Problem:
Nvim fails to create tempfile "…/nvim6UJx04/7" when 'shell' is set to
pwsh (PowerShell Core). This breaks filtered shell commands ":{range}!".
With shell set to cmd, it works.

Solution:
PowerShell doesn't use "<" for stdin redirection. Instead, use
"-RedirectStandardInput".

Closes #15913

3N4N commented:
Oh, I forgot to follow the editorconfig guidelines. Should I fix that and add another commit?

3N4N commented:
I actually haven't followed any of the contribution guidelines. Don't know how to run tests on Windows. I had a hard time just building nvim on Windows. And some of the other steps, like CI stuff, are out of my league of expertise.

justinmk commented:
@3N4N thanks for looking into this! To exercise this, it should be fairly straightforward to write a test that checks Nvim's `:verbose` log. So the test can check the actual shell command that is produced. 

Example: https://github.com/neovim/neovim/blob/f876a0eeb5af6f41acb187eb9442062c3984b26d/test/functional/vimscript/system_spec.lua#L220

3N4N commented:
@justinmk I'm sorry. I have no idea how to write tests. Plus, I'm on Windows, and I don't know how to run tests here. I built neovim with [Mingw-w64](https://github.com/neovim/neovim/wiki/Building-Neovim#windows--msys2--mingw).

3N4N commented:
@zeertzjq sorry for tagging you, but can I get some help getting this PR merged? It'd be really cool to help in the dev, being a neovim fan, you know.

3N4N commented:
@justinmk I wrote a test. Please see if it'll do. And I'm the one whom you answered about testing with `screen:expect` in Matrix chat.

justinmk commented on test/functional/vimscript/system_spec.lua:
```diff
@@ -190,6 +190,27 @@ describe('system()', function()
+        nvim('set_option', 'shellredir', '-RedirectStandardOutput %s -NoNewWindow -Wait')
+        nvim('set_option', 'shellpipe', '-RedirectStandardOutput %s -NoNewWindow -Wait')
+        nvim('set_option', 'shellquote', '')
+        nvim('set_option', 'shellxquote', '')
```
can use `helpers.set_shell_powershell()` instead (which also needs to be updated--all other powershell tests will use it)

justinmk commented on test/functional/vimscript/system_spec.lua:
```diff
@@ -190,6 +190,27 @@ describe('system()', function()
       eq(iswin() and '??\n' or 'ああ\n', eval([[system('Write-Output "ああ"')]]))
     end)
 
+    it('check powershell filter and redirection (#16271)', function()
```
since there is no unix test here, let's add it to the windows-only tests here:

https://github.com/neovim/neovim/blob/e420cd6c67a0aad5b5dcb40748f733876e66a2c3/test/functional/vimscript/system_spec.lua#L128

3N4N replied:
So, the test should be run for Unix, too. So I moved it out of `if iswin()`.

justinmk commented on test/functional/vimscript/system_spec.lua:
```diff
@@ -190,6 +190,27 @@ describe('system()', function()
+        feed('i3<cr>1<cr>2<esc>:write<cr>')
+        feed(':4verbose %!sort<cr>')
+        screen:expect{
+          any=[[3 lines filtered]]
```
can we also check that the args (RedirectStandardOutput, etc) are printed ? That's why `:4verbose` is used, right?

3N4N commented:
@justinmk I resolved the the reviews. I ended up resizing the screen to 500 and checking for the args with pattern matching (".*"). If someone builds nvim in a folder with more than about 500 characters, the check will fail. But this is what I could come up with.

Another thing. Check this code block.

```c
char *make_filter_cmd(char *cmd, char *itmp, char *otmp)
{
  bool is_fish_shell =
#if defined(UNIX)
    STRNCMP(invocation_path_tail(p_sh, NULL), "fish", 4) == 0;
#else
    false;
#endif
  bool is_pwsh =
#if defined(UNIX)
    false;
#else
    STRNCMP(invocation_path_tail(p_sh, NULL), "pwsh", 4) == 0
    || STRNCMP(invocation_path_tail(p_sh, NULL), "powershell", 10) == 0;
#endif
```

Imitating the code-block for `is_fish_shell`, if in a Unix machine, I blindly assumed the shell isn't powershell. But powershell, in the form of pwsh, is available for Unix machines, too. I should remove that if-unix block, right?

justinmk commented:
This is looking great, thank you! Just a few minor things then we can merge it.

> But powershell, in the form of pwsh, is available for Unix machines, too. I should remove that if-unix block, right?

yes

justinmk commented on test/functional/vimscript/system_spec.lua:
```diff
@@ -173,6 +173,23 @@ describe('system()', function()
+        feed(':edit ' .. tempfile .. '<cr>')
+        feed('i3<cr>1<cr>2<esc>:write<cr>')
+        feed(':4verbose %!sort<cr>')
+        screen:expect{
+          any=[[3 lines filtered]]
+        }
+        screen:expect{
+          unchanged=true,
+          any=[[Executing command: "Start%-Process sort %-RedirectStandardInput .* %-RedirectStandardOutput .* %-NoNewWindow %-Wait]]
+        }
```
should be able to avoid the `unchanged=true` by combining these:

```suggestion
        screen:expect{
          any=[[3 lines filtered.*Executing command: "Start%-Process sort %-RedirectStandardInput .* %-RedirectStandardOutput .* %-NoNewWindow %-Wait]]
        }
```

3N4N replied:
Done.

3N4N commented:
Any update on this? Should I look into why these tests are failing? It might be because this branch is built atop an older commit. Should I rebase on master or release-0.7?

clason commented:
Yes, test failures are relevant and need to be looked into. (I don't know anything about Windows, so can't help with the actual changes.)

You can rebase on master, if you want, but it should not be necessary. (Don't rebase on `release-0.7`, commits are cherry-picked into this branch after merging to master if they are applicable to a point release.)

3N4N commented:
Turns out the `shellpipe` needs to use `Out-File` and `shellredir` needs to use `-RedirectStandardOutput`. I updated the tests. The errors should be fixed now.

I also updated the `make_filter_cmd` function to accommodate pwsh in Linux. Working on both my personal Windows and Ubuntu installations. Let's see how the tests turn out.

3N4N commented:
@clason would you check this error, please? It's bugging out here, but my commit should have nothing to do with this. Also, it's not Windows-related/specific.

https://github.com/neovim/neovim/blob/6d52a29c3b8714804facdc7705e57e7f0511c85a/test/functional/plugin/health_spec.lua#L35-L40

zeertzjq commented on src/nvim/ex_cmds.c:
```diff
@@ -1598,13 +1605,24 @@ char *make_filter_cmd(char *cmd, char *itmp, char *otmp)
-    xstrlcat(buf, (const char *)itmp, len - 1);
+    // xstrlcat(buf, " < ", len - 1);
+    // xstrlcat(buf, (const char *)itmp, len - 1);
+    if (is_pwsh) {
+      xstrlcat(buf, " -RedirectStandardInput ", len);
+    } else {
+      xstrlcat(buf, " < ", len);
+    }
+    xstrlcat(buf, (const char *)itmp, len);
```
Can it be caused by the change from `len - 1` to `len` here?

This `len - 1` seem to be need for to make room for the final NUL byte: <https://github.com/neovim/neovim/commit/d97d66e173b11a26015a99bed8467081d9c8733e>

zeertzjq replied:
```suggestion
    if (is_pwsh) {
      xstrlcat(buf, " -RedirectStandardInput ", len - 1);
    } else {
      xstrlcat(buf, " < ", len - 1);
    }
    xstrlcat(buf, (const char *)itmp, len - 1);
```

3N4N replied:
I'll edit it and push again. Thanks! for finding that.

3N4N commented:
It's green! It's ALL green! :laughing: 

I'll fixup the redundant codes (commented codes which were for my benefit) and then push one last time.

3N4N commented:
```
[ RUN      ] treesitter parser API supports caching queries: 413.68 ms FAIL
test/functional/treesitter/parser_spec.lua:188: firstrun: 384900 ms, manyruns: 882 ms
```

For this test:
https://github.com/neovim/neovim/blob/6d52a29c3b8714804facdc7705e57e7f0511c85a/test/functional/treesitter/parser_spec.lua#L170-L189

3N4N commented:
Update?

zeertzjq commented on test/functional/vimscript/system_spec.lua:
```diff
@@ -190,6 +190,23 @@ describe('system()', function()
+        pending("not tested; powershell was not found", function() end)
+        return
+      end
+      local tempfile = helpers.tmpname()
+      helpers.set_shell_powershell()
+      screen:try_resize(500, 8)
+      feed(':edit ' .. tempfile .. '<cr>')
+      feed('i3<cr>1<cr>2<esc>:write<cr>')
```
This will leave behind a `tempfile`. Is writing to a file necessary in this test?

3N4N replied:
It's not necessary. Technically I can call `:%!sort` on an empty file. But I'm running into a problem here. For some reason, that same test, just without ".*3 lines filtered" is causing the running of test to get stuck (I have to ctrl-c out of it).

```lua
    it('works with powershell for filter and redirection (#16271)', function()
      -- same thing as before
      screen:expect{
        any=[[Executing command: "Start%-Process sort %-RedirectStandardInput .* %-RedirectStandardOutput .* %-NoNewWindow %-Wait]]
      }
      feed('<cr>')
    end)
```

The output gets stuck here:

```
enan@cirdan:~/projects/neovim$ make functionaltest TEST_FILE=test/functional/vimscript/system_spec.lua
mkdir -p build
touch build/.ran-third-party-cmake
ninja  -C .deps
ninja: Entering directory `.deps'
[3/3] cd /home/enan/projects/neovim/.deps && /usr/bin/cmake -E touch .third-party
ninja  -C build
ninja: Entering directory `build'
[0/2] Re-checking globbed directories...
[1/6] cd /home/enan/projects/neovim && /home/...build/config/auto/versiondef_git.h v0.8.0-dev
ninja  -C build functionaltest
ninja: Entering directory `build'
[0/2] Re-checking globbed directories...
[1/2] cd /home/enan/projects/neovim/build && ...ome/enan/projects/neovim/cmake/RunTests.cmake
[----------] Global test environment setup.
[----------] Running tests from test/functional/vimscript/system_spec.lua
[ RUN      ] system() command passed as a List throws error if cmd[0] is not executable: 3.61 ms OK
[ RUN      ] system() command passed as a List parameter validation does NOT modify v:shell_error: 5.28 ms OK
[ RUN      ] system() command passed as a List quotes arguments correctly #5280: 5.53 ms OK
[ RUN      ] system() command passed as a List calls executable in $PATH: 3.61 ms SKIP
[ RUN      ] system() command passed as a List does NOT run in shell: 4.20 ms OK
[ RUN      ] system() sets v:shell_error: 10.16 ms OK
[ RUN      ] system() executes shell function works with powershell w/ UTF-8 text (#13713): 518.15 ms OK
[ RUN      ] system() executes shell function works with powershell for filter and redirection (#16271): 11.12 ms OK
```
So this test runs okay, but then the rest gets stuck. I must be misunderstanding how screen:expect works.

3N4N replied:
Not necessary. I can call `:%!sort` on an empty file (or `[No Name]` file). I've pushed the changed test.

clason commented:
Usual flake on those Github runners. Just ignore it and re-run if needed.

justinmk commented on test/functional/ex_cmds/shell_filter_spec.lua:
```diff
@@ -0,0 +1,35 @@
+
+before_each(clear)
+
+describe('shell filtering', function()
```
don't want a separate test file for this. why move it?

zeertzjq replied:
I just think this is unrelated to `system()`

justinmk replied:
ok, we could rename  system_spec.lua to shell_spec and maybe add a describe('shell') group. I have some local changes, so please avoid pushing to the PR.

justinmk committed:
refactor(tests): introduce testprg()

Also:
- Add a describe('shell :!') section to system_spec.
- Make the test for #16271 work on systems without powershell.

justinmk commented:
pushed an update. got a bit carried away with cleaning up some old debt in the tests, but the relevant change is:

- the test now runs on all systems by using a fake `pwsh-test` binary. Might be useful in other tests too, but I'll save that for later

justinmk commented on src/nvim/ex_cmds.c:
```diff
@@ -1590,7 +1594,10 @@ char *make_filter_cmd(char *cmd, char *itmp, char *otmp)
 #if defined(UNIX)
```
TODO: there's a fair amount of duplicate code in these defines

justinmk commented on runtime/doc/options.txt:
```diff
@@ -5292,9 +5292,9 @@ A jump table for the options with a short description can be found at |Q_op|.
 		let &shellcmdflag = '-NoLogo -NoProfile -ExecutionPolicy RemoteSigned -Command [Console]::InputEncoding=[Console]::OutputEncoding=[System.Text.Encoding]::UTF8;'
-		let &shellredir = '2>&1 | Out-File -Encoding UTF8 %s; exit $LastExitCode'
+		let &shellredir = '-RedirectStandardOutput %s -NoNewWindow -Wait'
 		let &shellpipe = '2>&1 | Out-File -Encoding UTF8 %s; exit $LastExitCode'
```
If `>` doesn't work with powershell then why is it used for 'shellpipe'?

3N4N replied:
There are two cases:

1. `Start-Process sort -RedirectStandardInput filein.txt -RedirectStandardOutput fileout.txt`: requires redirection.
2. `Start-Process dir`: does not require redirection

In the second case, if the command `dir` does not exist, that is, if the user provides a wrong command, nvim needs to write the error to a file (at least that's my understanding, from [tests like this](https://github.com/neovim/neovim/blob/b7084fef4c850d0352488b14dcff0f36a7e75e1c/test/functional/ex_cmds/make_spec.lua#L24-L31)). Now, pwsh doesn't let users direct both std in and err to the same file, that is, it doesn't let us use something like `2>&1`. If you give the same file to both `-RedirectStandardOutput` and `RedirectStandardError`, pwsh will give error. Stackoverflow suggested simply using `Out-File` to redirect everything to a file.

But `Out-File` doesn't work with case 1. My understanding is that, this use case is why vim has two different options for redirection and piping. But I don't have any confirmation except that it seemed to work.

justinmk commented on runtime/doc/options.txt:
```diff
@@ -5292,9 +5292,9 @@ A jump table for the options with a short description can be found at |Q_op|.
+		let &shell = executable('pwsh') ? 'pwsh' : 'powershell'
 		let &shellcmdflag = '-NoLogo -NoProfile -ExecutionPolicy RemoteSigned -Command [Console]::InputEncoding=[Console]::OutputEncoding=[System.Text.Encoding]::UTF8;'
-		let &shellredir = '2>&1 | Out-File -Encoding UTF8 %s; exit $LastExitCode'
+		let &shellredir = '-RedirectStandardOutput %s -NoNewWindow -Wait'
```
Since `exit $LastExitCode` was removed, will Nvim correctly get the exit code after this change?

3N4N replied:
I don't know that. But running commands with wrong arguments in cmd directly gives the following results:

```cmd
C:\Users\ACER> pwsh -c "sldkfj  2>&1 | Out-File -Encoding UTF8 err.txt; exit $LastExitCode"
sldkfj: The term 'sldkfj' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

C:\Users\ACER> echo Exit Code is %errorlevel%
Exit Code is 0

C:\Users\ACER> pwsh -c "sldkfj  2>&1 | Out-File -Encoding UTF8 err.txt"
sldkfj: The term 'sldkfj' is not recognized as a name of a cmdlet, function, script file, or executable program.
Check the spelling of the name, or if a path was included, verify that the path is correct and try again.

C:\Users\ACER> echo Exit Code is %errorlevel%
Exit Code is 1

C:\Users\ACER> pwsh -c "Start-Process sort -RedirectStandardInput sldkfj.txt -RedirectStandardOutput out.txt -NoNewWindow -Wait"
Start-Process: This command cannot be run because either the parameter "RedirectStandardInput 'C:\Users\ACER\sldkfj.txt'" has a value that is not valid or cannot be used with this command. Give a valid input and Run your command again.

C:\Users\ACER> echo Exit Code is %errorlevel%
Exit Code is 1

C:\Users\ACER> pwsh -c "Start-Process sort -RedirectStandardInput sldkfj.txt -RedirectStandardOutput out.txt -NoNewWindow -Wait; exit $LastExitCode"
Start-Process: This command cannot be run because either the parameter "RedirectStandardInput 'C:\Users\ACER\sldkfj.txt'" has a value that is not valid or cannot be used with this command. Give a valid input and Run your command again.

C:\Users\ACER> echo Exit Code is %errorlevel%
Exit Code is 0
```

So, `; exit $LastExitCode` spends (?) the exit code? I don't know how Nvim handles that. I'm sure you can understand whether `; exit $LastExitCode` is needed or not from the above examples.

3N4N commented:
Holy. Thanks for the merge. And thanks for reviewing my commit, @justinmk @zeertzjq
